model/plugin_kvset_options.go

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package model
+
+import (
+	"encoding/json"
+	"net/http"
+)
+
+type PluginKVSetOptions struct {
+	EncodeJSON      bool
+	Atomic          bool
+	OldValue        interface{}
+	ExpiryInSeconds int64
+}
+
+func (opt *PluginKVSetOptions) IsValid() *AppError {
+	if !opt.Atomic && opt.OldValue != nil {
+		return NewAppError("PluginKVSetOptions.IsValid", "model.plugin_kvset_options.is_valid.old_value.app_error", map[string]interface{}{"Atomic": opt.Atomic, "OldValue": opt.OldValue}, "OldValue should not be defined on non atomic sets", http.StatusBadRequest)
+	}
+
+	return nil
+}
+
+func (opt *PluginKVSetOptions) GetPluginKeyValue(pluginId string, key string, value interface{}) (*PluginKeyValue, *AppError) {
+	serializedValue, err := opt.serializeValue(value)
+	if err != nil {
+		return nil, err
+	}
+
+	expireAt := int64(0)
+	if opt.ExpiryInSeconds > 0 {
+		expireAt = GetMillis() + (opt.ExpiryInSeconds * 1000)
+	}
+
+	kv := &PluginKeyValue{
+		PluginId: pluginId,
+		Key:      key,
+		Value:    serializedValue,
+		ExpireAt: expireAt,
+	}
+
+	return kv, nil
+}
+
+func (opt *PluginKVSetOptions) GetOldValueSerialized() ([]byte, *AppError) {
+	return opt.serializeValue(opt.OldValue)
+}
+
+func (opt *PluginKVSetOptions) serializeValue(value interface{}) ([]byte, *AppError) {
```

Comment:
I want this method to be internal to the struct... as far as I've read, that's not really a possibility in Go. Should I even care?

Reply:
No. Golang only allows exported or non-exported at package boundaries. Don't worry about it.